src/docs/design-patterns/decorator-pattern.md

```diff
@@ -0,0 +1,214 @@
+
+- **Adding Responsibilities Dynamically**: Use the Decorator pattern when you need to add additional responsibilities to objects at runtime without affecting other objects.
+
+- **Modular and Reusable Extensions**: When you want to create functionalities that can be applied to objects without changing those objects' classes, making the extensions modular and reusable across different scenarios.
+
+- **Avoiding Subclass Explosion**: If extending functionality through subclassing leads to an excessive number of classes, the Decorator pattern can help manage this complexity by allowing you to combine behaviors.
+
+- **Extension Without Modification**: In situations where modifying the original class is not possible due to restrictions like access levels or when the class is part of a library you do not control, decorators provide a way to add behavior.
+
+- **Conditional Behavior Addition**: When the behavior needs to be added to objects only under certain conditions or contexts, decorators can be dynamically applied and removed.
+
+- **Enhancing Legacy Code**: For enhancing or fixing legacy code where changing the original codebase is risky or impractical, applying decorators can introduce new behavior or fix issues without touching the original code.
+
+- **Compliance and Logging**: In scenarios requiring compliance, logging, or audit trails where these concerns are secondary to the main business logic, decorators can inject these cross-cutting concerns transparently.
+
+- **Validation and Security**: To add validation, authentication, or other security layers around core functionalities without embedding these concerns directly into the primary classes.
+
+- **Flexible Alternative to Inheritance**: When inheritance is too rigid or not suitable for adding functionalities, the Decorator pattern offers a more flexible solution by allowing mixing and matching of behaviors.
+
+Utilizing the Decorator pattern under these circumstances can lead to cleaner, more maintainable, and scalable code architectures by promoting separation of concerns and adherence to the Open/Closed Principle.
+
+## C# Example
+
+Let's implement the Decorator Pattern in a .NET application. We'll start with a basic `BookService` that adds a book to a repository. Then, we'll add decorators for logging with a stopwatch/timer and for validating the book using [FluentValidation](https://docs.fluentvalidation.net/en/latest/).
+
+### The Underlying Service
+
+First, define the `IBookService` interface and implement a basic `BookService`.
+
+```csharp
+public interface IBookService
+{
+    void AddBook(Book book);
+}
+
+public class BookService : IBookService
+{
+    private readonly IRepository<Book> _repository;
+
+    public BookService(IRepository<Book> repository)
+    {
+        _repository = repository;
+    }
+
+    public void AddBook(Book book)
+    {
+        _repository.Add(book);
+    }
```

Comment:
Should this be `IBookRepository`? You refer to `IBookRepository` in the DI part.